### PR TITLE
feat: rust binary search utilities

### LINKED ISSUE #471: Binary search utilities in Rust
We have binary search utilities in Python, but not in Rust. It would be ideal to have symmetry in our interfaces across languages.

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #471

### CODE REVIEW

**Comment by Shoeboxam** [comment]
* **#472** <a href="https://app.graphite.com/github/pr/opendp/opendp/472?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/opendp/opendp/472?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: rust/src/utilities/search/mod.rs [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+macro_rules! impl_binary_searchable_int {
+    ($($ty:ty),+ $(,)?) => {
+        $(impl BinarySearchable for $ty {
+            fn midpoint(lower: &Self, upper: &Self) -> Self {
+                if lower < &<$ty>::zero() && upper >= &<$ty>::zero() {
+                    (lower + upper).halve()
+                } else {
+                    lower + (upper - lower).halve()
+                }
+            }
```
Comment (mccalluc) [inline rust/src/utilities/search/mod.rs:42]: Add a comment to explain why this is necessary.

File: rust/src/utilities/search/mod.rs [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    signed_binary_search(predicate, bounds).map(|(value, _sign)| value)
+}
+
+/// Like [`binary_search`], but also returns the direction away from the decision boundary.
+///
+/// A returned sign of `1` means the passing side is above the boundary, and `-1` means it is below.
+pub fn signed_binary_search<T>(
+    predicate: impl Fn(&T) -> bool,
+    bounds: Option<(T, T)>,
+) -> Fallible<(T, i8)>
```
Comment (mccalluc) [inline rust/src/utilities/search/mod.rs:78]: How much parallelism should I expect between this and the python version? 

In python, there's an extra parameter `return_sign` to indicate whether the result includes this information: I think having a separate function like you have here is a better interface: The variability in the return type in python adds complexity.

Perhaps a follow-up issue to bring that in line with this?

File: rust/src/utilities/search/mod.rs [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+                if sign > 0 {
+                    if let Some(next) = center.checked_add(1) {
+                        bands.push(next);
+                    }
+                } else if let Some(next) = center.checked_sub(1) {
+                    bands.push(next);
+                }
+
+                for k in 1..=8 {
+                    let offset = match <$ty>::try_from(65_536_i128 * k as i128) {
```
Comment (mccalluc) [inline rust/src/utilities/search/mod.rs]: Replace with power of 2, to avoid typos?

File: rust/src/utilities/search/mod.rs [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+                    return Vec::new();
+                }
+
+                let mut bands = vec![center];
+                if let Some(next) = center.checked_add(1) {
+                    bands.push(next);
+                }
+
+                for k in 1..=8 {
+                    let offset = match <$ty>::try_from(65_536_u128 * k as u128) {
```
Comment (mccalluc) [inline rust/src/utilities/search/mod.rs]: ditto.

File: rust/src/utilities/search/test.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
+use super::*;
+
+#[test]
+fn test_binary_search_matches_python_r_int_behavior() -> Fallible<()> {
```
Comment (mccalluc) [inline rust/src/utilities/search/test.rs]: What's the "r"?

File: rust/src/utilities/search/test.rs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    assert_eq!(binary_search::<i32>(|x| *x <= -5, None)?, -5);
+    assert_eq!(binary_search::<i32>(|x| *x <= 5, None)?, 5);
+    assert_eq!(binary_search::<i32>(|x| *x >= -5, None)?, -5);
+    assert_eq!(binary_search::<i32>(|x| *x >= 5, None)?, 5);
+    assert_eq!(binary_search::<u32>(|x| *x > 5, None)?, 6);
```
Comment (mccalluc) [inline rust/src/utilities/search/test.rs]: This last isn't one of the python tests. Add to python tests, or explain why rust is different?

File: rust/src/utilities/search/test.rs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+fn test_exponential_bounds_search_matches_python_r_bands() {
+    assert_eq!(
+        exponential_bounds_search::<i32>(&|x| *x > 5),
+        Some((1, 65_536))
+    );
+    assert_eq!(
+        exponential_bounds_search::<f64>(&|x| *x > 5.0),
+        Some((2.0, 16.0))
+    );
+}
```
Comment (mccalluc) [inline rust/src/utilities/search/test.rs:43]: I don't see a corresponding python test?

**Review by mccalluc — COMMENTED** [review]
Some comments, but they are superficial. Bigger questions:
- Why now? Is something you are planning to use soon?
- Since this will be in the rust core, can you remove the python implementation? (Several of the comments are about python-rust parallels, and if we don't have a duplicate implementation in python, those go away.)
- Can the utilities also be generated for R?
- This feels like a really good target for https://hypothesis.readthedocs.io/ style tests. Probably not in this PR, but it just feels like there is lots of details in the code, and we can't anticipate the edge cases.

File: python/src/opendp/mod.py [hunk 8]
```diff
@@ -1443,6 +1443,47 @@ def binary_search_param(
 
     return binary_search(lambda param: make_chain(param).check(d_in, d_out), bounds, T)
 
+
+def _call_rust_search(
+    predicate: Callable[[float], bool],
+    T: "RuntimeType | str",
+    search: Callable[[Callable[[float], bool]], Any],
+    bounds: tuple[float, float] | None = None,
+) -> Any:
```
Comment (mccalluc) [inline python/src/opendp/mod.py:1452]: A little surprised by the two `Any`: consider using an `assert` to constrain the type returned by rust?

File: python/src/opendp/mod.py [hunk 9]
```diff
@@ -1443,6 +1443,47 @@ def binary_search_param(
...
+        nonlocal first_exception, had_success
+        try:
+            result = predicate(arg)
+        except Exception as err:
+            if first_exception is None:
+                first_exception = err
+            raise
+        else:
+            had_success = True
+            return result
```
Comment (mccalluc) [inline python/src/opendp/mod.py:1466]:
```suggestion
        had_success = True
        return result
```
Equivalent? but maybe the `else` is more clear?

File: python/test/test_binary_search.py [hunk 10]
```diff
@@ -76,7 +77,36 @@ def predicate(_):
...
+
+    with pytest.raises(ValueError, match="predicate is constant"):
+        _call_rust_search(lambda _: True, "i32", search)
+
+
+def test_binary_search_forwards_untranslated_rust_errors():
+    def search(_):
+        raise OpenDPException("FailedFunction", "callback failed")
+
+    with pytest.raises(OpenDPException, match='FailedFunction\\("callback failed"\\)'):
```
Comment (mccalluc) [inline python/test/test_binary_search.py:101]: consider `re.escape` or
```suggestion
    with pytest.raises(OpenDPException, match=r'FailedFunction\("callback failed"\)'):
```

**Review by mccalluc — APPROVED** [review]
Thanks. Looks good. A couple questions, but they shouldn't block.
